### PR TITLE
perf: vectorize signal/rejected candle export in backtesting reports

### LINKED ISSUE #13318: Backtesting --export signals post-processing is quadratic and can take hours for signal-dense strategies
## Describe your environment

  * Operating system: Ubuntu 24.04 (WSL2)
  * Python Version: 3.12.3
  * CCXT version: 4.5.61
  * Freqtrade Version: 2026.6 (also verified unchanged on current `develop`, cf5280f)

## Describe the problem:

When backtesting with `--export signals`, the post-processing step that collects signal/rejected candles (`generate_trade_signal_candles` / `generate_rejected_signals` in `freqtrade/optimize/optimize_reports/optimize_reports.py`) is quadratic in the number of trades / rejected signals:

```python
for t in signals:
    data_df_row = pairdf.loc[(pairdf["date"] == t[0])].copy()   # full-frame scan per signal
    ...
    rejected_signals_only_df = concat(                           # re-concat of the whole
        [rejected_signals_only_df.infer_objects(), data_df_row.infer_objects()]  # accumulated frame per signal
    )
```

Each rejected signal triggers (a) a full boolean scan of the pair dataframe and (b) a `concat` that copies the entire accumulated result again. `generate_trade_signal_candles` has the same pattern (`pairdf.loc[pairdf["date"] < v[date_col]]` + `concat` per trade row).

For strategies that emit entry signals on many candles (grid-like / always-scoring strategies, where most signals get rejected by `max_open_trades`), a multi-year backtest accumulates tens of thousands of rejected signals per pair, and this step runs for **hours** while the backtest loop itself finishes in minutes. From the outside it looks like the backtest hangs at 100% CPU with no output, since the loop has already completed and nothing is written until export finishes.

### Measured impact (real workload)

19 pairs, 15m, `20230101-20260101`, a signal-dense grid strategy, `max_open_trades 2`:

* `--export none`: backtest completes in ~2 minutes.
* `--export signals`: killed after **66+ minutes**, still inside `generate_rejected_signals`.

`py-spy dump` of the "hanging" process:

```text
Thread (active): "MainThread"
    _concat_same_type (pandas/core/arrays/arrow/array.py:1847)
    concat_compat (pandas/core/dtypes/concat.py:83)
    concatenate_managers (pandas/core/internals/concat.py:136)
    _get_result (pandas/core/reshape/concat.py:657)
    concat (pandas/core/reshape/concat.py:466)
    generate_rejected_signals (freqtrade/optimize/optimize_reports/optimize_reports.py:161)
    backtest_one_strategy (freqtrade/optimize/backtesting.py:1855)
    start (freqtrade/optimize/backtesting.py:1913)
```

Synthetic micro-benchmark (105k candles, single pair): 2,000 rejected signals already take ~7.2s with the current implementation; scaling is quadratic (20,000 signals extrapolate to ~12 minutes *per pair*), while a vectorized `merge` handles 20,000 signals in ~0.02s.

## Steps to reproduce:

1. Take any strategy that sets `enter_long` on most candles (so that `max_open_trades` rejects most signals).
2. Run a multi-year backtest with `--export signals` on a 15m timeframe.
3. Observe the process staying at 100% CPU long after `Backtesting with data from ... up to ...` finished, with the time spent in `generate_rejected_signals`.

## Proposed fix

Both functions can be vectorized with identical output (same rows, same order, same index):

* `generate_rejected_signals`: build a small DataFrame from the rejected `(date, enter_tag)` tuples and `merge` it with the pair dataframe on `date` (one pass instead of k scans + k concats).
* `generate_trade_signal_candles`: use `Series.searchsorted(..., side="left") - 1` to find the last candle strictly before each trade date, then a single `iloc` take.

I have a branch with this change, verified row-identical against the current implementation (including duplicate dates, empty-signal pairs and empty-result pairs) plus unit tests for both functions — PR follows.

### PR DESCRIPTION
**Pull request by halmisen**
## Summary

Solves #13318.

`generate_trade_signal_candles()` and `generate_rejected_signals()` (used by `--export signals`) built their results with a full pair-dataframe scan **plus** a re-`concat` of the whole accumulated frame *per trade / per rejected signal* — O(k²) in the number of trades/rejected signals. For signal-dense strategies over multi-year timeranges this post-processing step runs for hours after the backtest loop has already finished (details, py-spy stack and measurements in the issue).

This PR vectorizes both functions:

* `generate_rejected_signals`: one `merge` of the pair dataframe with a small `(date, enter_tag)` frame, instead of k scans + k concats.
* `generate_trade_signal_candles`: `Series.searchsorted(..., side="left") - 1` to pick the last candle strictly before each trade date, then a single `iloc` take, instead of a `<`-mask scan + concat per trade.

Output is unchanged: same rows, same row order, same index, duplicate trade/signal dates preserved, empty-signal and empty-result pairs behave as before. Verified row-identical against the previous implementations on synthetic data covering these cases (plus an injected-difference canary to prove the comparison can fail), and benchmarked at 105k candles: 20k rejected signals go from ~12 min (extrapolated quadratic) to ~0.02 s per pair; on the real 19-pair / 3-year workload from the issue, `--export signals` goes from 66+ min (killed, unfinished) to seconds of export overhead.

Also adds unit tests for both functions (previously untested) covering duplicate dates, empty signal lists and pairs without trades.

## Quick changelog

- perf: vectorize `generate_trade_signal_candles` / `generate_rejected_signals` (O(k²) → O(n+k)) in backtesting `--export signals` post-processing
- tests: add unit tests for both functions

### CODE REVIEW

File: freqtrade/optimize/optimize_reports/optimize_reports.py [hunk 1]
```diff
@@ -150,19 +146,17 @@ def generate_rejected_signals(
...
-            data_df_row["pair"] = pair
-            data_df_row["enter_tag"] = t[1]
-
-            rejected_signals_only_df = concat(
-                [rejected_signals_only_df.infer_objects(), data_df_row.infer_objects()]
-            )
-
-        rejected_candles_only[pair] = rejected_signals_only_df
+        signals_df = DataFrame(signals, columns=["date", "enter_tag"])
+        rejected = pairdf.merge(signals_df, on="date", how="inner").infer_objects()
```
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports/optimize_reports.py:159]: You can get a `KeyError: "['enter_tag'] not in index"  ` if you `populate_indicators()`: assigns enter_tag.
(e.g. `dataframe["enter_tag"] = "whatever"`).

While not common - it's also not really forbidden to assign enter_tag as part of populate_indicators. This was (silently) covered by previous overwriting - but will cause a name collision (and as such - `enter_tag_x` and `enter_tag_y` on the merge.

File: freqtrade/optimize/optimize_reports/optimize_reports.py [hunk 2]
```diff
@@ -127,21 +127,17 @@ def generate_trade_signal_candles(
...
-                    [signal_candles_only_df.infer_objects(), signal_inds.infer_objects()]
-                )
-
-            signal_candles_only[pair] = signal_candles_only_df
+            if pairresults.empty:
+                signal_candles_only[pair] = DataFrame()
+                continue
+            # Last candle with date strictly before the trade date, for each trade.
+            candle_idx = pairdf["date"].searchsorted(pairresults[date_col], side="left") - 1
+            signal_candles_only[pair] = pairdf.iloc[candle_idx].infer_objects()
```
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports/optimize_reports.py:142]: we might want an additional check for negative indexes here (that's a silent potential regression - which is now silent instead of "loud")

while it shouldn't happen - a negative value here would add the last row to `signal_candles_only` - instead of the first one.

**Review by xmatthias — CHANGES_REQUESTED** [review]
some minor nitpicks / safety checks - but overall, this looks great  and accomplishes the promised time improvement 👍

**Review by xmatthias — APPROVED** [review]
Adressed the minor nitpicks myself today so we can get this enhancement added.

LGTM 👍
